tool/src/java/mod.rs

```diff
@@ -0,0 +1,963 @@
+                    r#"var {name}MemSeg = arena.allocateFrom({name}, StandardCharsets.UTF_8);
+var {name}Len = {name}MemSeg.byteSize();"#
+                )
+                .into(),
+                format!("{name}MemSeg, {name}Len - 1").into(),
+            ),
+            hir::Type::Slice(Slice::Primitive(_, p)) => {
+                let primitive_ty = match p {
+                    hir::PrimitiveType::Bool => "JAVA_BOOLEAN",
+                    hir::PrimitiveType::Char => "JAVA_INT",
+                    hir::PrimitiveType::Byte => "JAVA_BYTE",
+                    hir::PrimitiveType::Int(IntType::I8 | IntType::U8) => "JAVA_BYTE",
+                    hir::PrimitiveType::Int(IntType::I16 | IntType::U16) => "JAVA_SHORT",
+                    hir::PrimitiveType::Int(IntType::I32 | IntType::U32) => "JAVA_INT",
+                    hir::PrimitiveType::Int(IntType::I64 | IntType::U64) => "JAVA_LONG",
+                    hir::PrimitiveType::IntSize(_) => "JAVA_LONG",
+                    hir::PrimitiveType::Int128(_) => {
+                        panic!("java backend doesn't support 128 bit integers")
+                    }
+                    hir::PrimitiveType::Float(hir::FloatType::F32) => "JAVA_FLOAT",
+                    hir::PrimitiveType::Float(hir::FloatType::F64) => "JAVA_DOUBLE",
+                };
+                (
+                    format!(
+                        r#"var {name}Len = {name}.length;
+var {name}MemSeg = arena.allocate({primitive_ty}, {name}Len);
+for (var i = 0; i < {name}Len; i++) {{
+    {name}MemSeg.setAtIndex({primitive_ty}, i, {name}[i]);
+}}"#
+                    )
+                    .into(),
+                    format!("{name}MemSeg, {name}Len").into(),
+                )
+            }
+            // hir::Type::Slice(Slice::Strs(_)) => ("".into(), "".into()),
+            x => panic!("Unexpected slice type {x:?}"),
+        };
+        ParamConversion {
+            converted_value,
+            conversion_def: conversion,
+        }
+    }
+
+    fn gen_slice_return_conversion(&self, ty: &Slice) -> Result<Cow<'cx, str>, String> {
+        let return_conversion: Cow<'cx, str> = match ty {
+            Slice::Str(None, _) | Slice::Primitive(None, _) => {
+                // We can probably do boxed returns by just relying on jna
+                // public double[] asBoxedSLice() {
+                //     try (var arena = Arena.ofConfined()) {
+                //         var nativeVal = somelib_h.Float64Vec_as_boxed_slice(arena, internal);
+                //         var data = dev.diplomattest.somelib.ntv.DiplomatF64View.data(nativeVal);
+                //         var len = dev.diplomattest.somelib.ntv.DiplomatF64View.len(nativeVal);
+                //         var returnVal = data.asSlice(0, len * JAVA_DOUBLE.byteSize()).toArray(JAVA_DOUBLE);
+                //         Native.free(data.address());
+                //         return returnVal;
+                //     }
+                // }
```

Comment:
Here's what a boxed slice would look like using jna. It should be the only place where we'd need to use it.